File: QuestionGenerator.py
import http.client
import json

def push_to_database(userId, classId, data):
    question, answer = data[0], data[1]
    conn = http.client.HTTPSConnection("nodepad-ecd25.firebaseio.com/.json")
    headers = {"Content-type": "application/json"}
    payload = dict()
    payload["userId"] = userId
    payload["classId"] = classId
    qa = dict()
    qa["Question"] = question
    qa["Answer"] = answer
    payload["QA"] = qa
    json_payload = json.dumps(payload)
    #with open("test.json") as test:
    conn.request("POST", "/markdown", json_payload, headers)
    response = conn.getresponse()
    print(response.read().decode())
    conn.close()

# Returns True if tag is a proper noun else False
def is_proper_noun(tag):
    return tag == "NNP" or tag == "NNPS"

# Returns a tuple (x, y) where x is the subject phrase and y is "Proper" if x is a proper noun else
# y is "Common"
def get_subject(tree):
    subject_type = "Common"
    first_noun_phrase_index = tree.find("NP")
    index = first_noun_phrase_index + 3
    char_array = list(tree)
    num_left_paren = 0
    num_right_paren = 0
    char_buffer = []
    subject = []
    while num_left_paren >= num_right_paren:
        if char_array[index] == "(":
            num_left_paren += 1
        elif char_array[index] == ")":
            num_right_paren += 1
            space_index = 0
            space_index = len(char_buffer)
            for i in reversed(range(len(char_buffer))):
                if char_buffer[i] == " ":
                    space_index = i
                    break
            word = "".join(char_buffer[space_index + 1:])
            if word:
                subject.append(word)
            word_type = "".join(char_buffer[:space_index + 1])
            if is_proper_noun(word_type.strip()):
                subject_type = "Proper"
            char_buffer = []
        else:
            char_buffer.append(char_array[index])
        index += 1
    return (" ".join(subject), subject_type)

def get_verb_phrase(tree):
    verb_phrase_index = tree.find("VP")
    index = verb_phrase_index + 3
    char_array = list(tree)
    num_left_paren = 0
    num_right_paren = 0
    char_buffer = []
    verb_phrase = []
    while num_left_paren >= num_right_paren:
        if char_array[index] == "(":
            num_left_paren += 1
        elif char_array[index] == ")":
            num_right_paren += 1
            space_index = len(char_buffer)
            for i in reversed(range(len(char_buffer))):
                if char_buffer[i] == " ":
                    space_index = i
                    break
            word = "".join(char_buffer[space_index + 1:])
            if word:
                verb_phrase.append(word)
            char_buffer = []
        else:
            char_buffer.append(char_array[index])
        index += 1
    return " ".join(verb_phrase)

#Attempts to generate a question starting with "When" using the raw text
def get_when_question_answer(note):
    during_index = note.find("during")
    before_index = note.find("before")
    after_index = note.find("after")
    in_index = note.find("in")
    on_index = note.find("on")

# Generates one question and answer based on one sentence.
# Raw is an array of the original sentence of the notes.
# Analyzed is an array of the part of speech corresponding to the word in raw.
def generate_question_answer(tree):
    question, answer = None, None
    #Subject type is "Proper" if subject has proper noun else "Common"
    subject, subject_type = get_subject(tree)
    verb_phrase = get_verb_phrase(tree)
    # Detect "Who" question
    if subject_type == "Proper":
        question = "Who " + verb_phrase + "?"
        answer = subject
        return (question, answer)
    elif subject_type == "Common":
        question = "What " + verb_phrase + "?"
        answer = subject
    return (question, answer)
    
note = "John Adams drafted the Declaration of Independence"
tree = "(TOP (S (NP (NNP John) (NNP Adams)) (VP (VBD wrote) (NP (NP (DT the) (NNP Declaration)) (PP (IN of) (NP (NNP Independence)))))))"
print(note)
print(generate_question_answer(tree))
note = "The mitochondria is the powerhouse of the cell"
tree = "(TOP (S (NP (DT The) (NN mitochrondria)) (VP (VBZ is) (NP (NP (DT the) (NN powerhouse)) (PP (IN of) (NP (DT the) (NN cell)))))))"
print(note)
print(generate_question_answer(tree))
note = "Declaration of Independence was drafted by John Adams"
tree = "(TOP (S (NP (NP (NNP Declaration)) (PP (IN of) (NP (NNP Independence)))) (VP (VBD was) (VP (VBN drafted) (PP (IN by) (NP (NNP John) (NNP Adams)))))))"
print(note)
print(generate_question_answer(tree))
push_to_database(1, 1, generate_question_answer(tree))

File: QuestionGenerator.js
// Setup Alchemy Language
var watson = require('watson-developer-cloud');
var alchemy_language = watson.alchemy_language({
  api_key: 'API-KEY'
})

// Generate random int
function getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

// Returns index of first noun phrase of treeArray
function getSlicingIndex(treeArray) {
    index = 0
    for (var i = 0; i < treeArray.length; i++) {
        var data = treeArray[i];
        data = data.slice(1, data.length - 1);
        var typeAndWord = data.split(" ");
        var type = typeAndWord[0];
        if (type.slice(0, 1) === "V") {
            index++;
        } else {
            break;
        }
    }
    return index;
}

// Turns original tree into readable array
function generateTreeArray(tree) {
    var pattern = /\([A-Z]* *[A-z]*\)/g;
    var treeArray = tree.match(pattern);
    return treeArray;
}

// Returns starting and ending vertices [x, y) of verb phrase
// Raw is array of raw words in sentence
// Tree is a string of the analyzed tree struction of the sentence
function getStartEndIndicesVerbPhrase(treeArray) {
    var startingVerbIndex = -1;
    var endingVerbIndex = -1;
    for (var i = 0; i < treeArray.length; i++) {
        var data = treeArray[i];
        data = data.slice(1, data.length - 1);
        var typeAndWord = data.split(" ");
        var type = typeAndWord[0];
        if (startingVerbIndex ===  -1) {
            if (type.slice(0, 1) === "V") {
                startingVerbIndex = i;
            }
        } else {
            if (type.slice(0, 1) != "V") {
                endingVerbIndex = i;
                break;
            }
        }
    }
    return [startingVerbIndex, endingVerbIndex];
}

// Returns String of noun phrase of raw sentece
function getNounPhrase(raw, stopIndex) {
    if (stopIndex == 0) {
        return null;
    }
    return raw.slice(0, stopIndex).join(" ");
}

// Returns String of verb phrase of raw sentence
function getVerbPhrase(raw, startIndex, stopIndex) {
    return raw.slice(startIndex, stopIndex).join(" ");
}

// Returns String of object phrase of raw sentence
function getObjectPhrase(raw, startIndex) {
    return raw.slice(startIndex).join(" ");
}

// Returns [x, y, z] where x is subject, y is verb, z is object phrase
// Note is the input sentence
// Tree is the result from Microsoft API
function generateMapping(note, tree) {
    var noteArray = note.split(" ");
    var treeArray = generateTreeArray(tree);
    var sliceIndex = getSlicingIndex(treeArray);
    noteArray = noteArray.slice(sliceIndex);
    treeArray = treeArray.slice(sliceIndex);
    var startEnd = getStartEndIndicesVerbPhrase(treeArray);
    var start = startEnd[0];
    var end = startEnd[1];
    var mapping = [getNounPhrase(noteArray, start), getVerbPhrase(noteArray, start, end), getObjectPhrase(noteArray, end)];
    return mapping;
}

// Pushes questions to Firebase
function pushQuestions(question, answer) {
    console.log("Q:", question, "A:", answer);
}

// Generates and pushes questions to Firebase
function generateAndPushQuestions(note, tree) {
    var mapping = generateMapping(note, tree);
    var blank = "_____";
    var parameters = {
        text: note
    };
    alchemy_language.relations(parameters, function (err, response) {
        if (err) {
            console.log('error:', err);
        } else {
            var relations = response["relations"];
            relations.forEach(function (relation) {
                if ("location" in relation) {
                    if ("subject" in relation && "action" in relation) {
                        var subject = relation["subject"]["text"].replace(/The/g, "the");
                        var action = relation["action"]["text"].replace(/was /g, "");
                        var question = "Where was " + subject + " " + action + "?";
                        var answer = relation["location"]["text"];
                        pushQuestions(question, answer);
                    }
                }
            })
        }
    });
    for (var i = 0; i < mapping.length; i++) {
        if (mapping[i] == "is" || mapping[i] == "was" || mapping[i] == "are" || mapping[i] == "were") {
            var value = getRandomInt(0, 1);
            if (value == 0) {
                question = "True/False: " + mapping.join(" ");
                answer = "True";
                console.log("Q:", question, "A:", answer);
            } else {
                question = "True/False " + mapping[0] + " " + mapping[1] + " not " + mapping[2];
                answer = "False";
                console.log("Q:", question, "A:", answer);
            }
            continue;
        }
        var temp = mapping[i];
        mapping[i] = blank;
        var question = mapping.join(" ");
        var answer = temp.replace(/[.,;]/g, "").replace(/\s{2,}/g," ");
        pushQuestions(question, answer)
        mapping[i] = temp;
    }
}

var note = "John Adams drafted the Declaration of Independence.";
var tree = "(TOP (S (NP (NNP John) (NNP Adams)) (VP (VBD drafted) (NP (NP (DT the) (NNP Declaration)) (PP (IN of) (NP (NNP Independence))))) (. .)))";
generateAndPushQuestions(note, tree);
note = "drafted the declaration of independence is a document";
tree = "(VP (VBD drafted) (NP (NP (DT the) (NN declaration)) (PP (IN of) (NP (NN independence)))) (VP (VBZ is) (NP (DT a) (NN document))))";
generateAndPushQuestions(note, tree);
note = "The Declaration of Independence was written in Pennsylvania.";
tree = "(TOP (S (NP (NP (DT The) (NNP Declaration)) (PP (IN of) (NP (NNP Independence)))) (VP (VBD was) (VP (VBN written) (PP (IN in) (NP (NNP Pennsylvania))))) (. .)))";
generateAndPushQuestions(note, tree);
